File: Week14/validSudoku36.java
public class Solution {
    public boolean isValidSudoku(char[][] board) {
        if(board == null || board.length != 9 || board[0].length != 9) {
            return false;
        }
        
        for(int i = 0; i < 9; i += 3) {
            for(int j = 0; j < 9; j += 3) {
                HashSet<Integer> sudokuGrid = new HashSet<Integer>();
                for(int m = i; m < i + 3; m++) {
                    for(int n = j; n < j + 3; n++) {
                        if(board[m][n] == '.') continue;
                        else if(board[m][n] - '0' > 9 || board[m][n] - '0' < 0 || sudokuGrid.contains(board[m][n] - '0')) {
                            return false;
                        } else {
                            sudokuGrid.add(board[m][n] - '0');
                        }
                    }
                }
            }
        }
        
        for(int i = 0; i < 9; i++) {
            HashSet<Integer> sudokuGrid = new HashSet<Integer>();
            for(int j = 0; j < 9; j++) {
                if(board[i][j] == '.') continue;
                else if(board[i][j] - '0' > 9 || board[i][j] - '0' < 0 || sudokuGrid.contains(board[i][j] - '0')) {
                    return false;
                } else {
                    sudokuGrid.add(board[i][j] - '0');
                }
            }
        }
        
        for(int j = 0; j < 9; j++) {
            HashSet<Integer> sudokuGrid = new HashSet<Integer>();
            for(int i = 0; i < 9; i++) {
                if(board[i][j] == '.') continue;
                else if(board[i][j] - '0' > 9 || board[i][j] - '0' < 0 || sudokuGrid.contains(board[i][j] - '0')) {
                    return false;
                } else {
                    sudokuGrid.add(board[i][j] - '0');
                }
            }
        }
        return true;
    }
}

File: Python/Week1/validSudoku36.py
class Solution(object):
    def isValidSudoku(self, board):
        """
        :type board: List[List[str]]
        :rtype: bool
        """
        if board == None or len(board) != 9 or len(board[0]) != 9:
            return False

        for row in xrange(3):
            for col in range(3):
                existed = [False] * 10
                for i in xrange(row * 3, row * 3 + 3):
                    for j in xrange(col * 3, col * 3 + 3):
                        entry = board[i][j]
                        if entry == '.':
                            continue
                        if existed[int(entry)]:
                            return False
                        else:
                            existed[int(entry)] = True

        for row in xrange(9):
            entry_list = board[row]
            existed = [False] * 10
            for entry in entry_list:
                if entry != '.':
                    if existed[int(entry)]:
                        return False
                    else:
                        existed[int(entry)] = True

        for col in xrange(9):
            existed = [False] * 10
            for row in xrange(9):
                entry = board[row][col]
                if entry != '.':
                    if existed[int(entry)]:
                        return False
                    else:
                        existed[int(entry)] = True

        return True


# xrange 与 range 的区别 range直接是list，xrange是list的前一步，循环很大时用xrange效果更好
